### PR TITLE
Multiple fixes related to agent IPAM plus an integration test.

### PR DESCRIPTION
**Pull request by jojimt**
Addresses duplicate IP addresses and other issues related to agent IPAM.

1. Race conditions due to lack of locking in the IPAM
2. Boundary conditions in the IPAM
3. RPC client close to free fds.

Added an integration test to exercise the interactions between CNI and the agent.

### CODE REVIEW

File: pkg/metadata/cnimetadata.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ type ContainerMetadata struct {
 
 func RecordMetadata(datadir string, network string, data ContainerMetadata) error {
 	dir := filepath.Join(datadir, network)
-	if err := os.MkdirAll(dir, 0644); err != nil {
+	if err := os.MkdirAll(dir, 0755); err != nil {
```
Comment (snaiksat) [inline pkg/metadata/cnimetadata.go]: Does this have any impact on upgrades? (as in, if the directory was created with 0644, and then we tried to read it below.)

Comment (jojimt) [reply pkg/metadata/cnimetadata.go]: No. This is a bug. I think the only reason this works is that the directory already exists with the right permission (perhaps created by the installer).
